File: 2024/day-4/part_1.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

const char string[] = "XMAS";

int findString(char[140][142], int, int);

int main(void) {
  FILE *file = fopen("input", "r");

  if (!file) {
    perror("File error");

    return EXIT_FAILURE;
  }

  char grid[140][142] = {0};
  int i = 0;

  while (fgets(grid[i++], 142, file))
    ;

  fclose(file);

  int answer = 0;

  for (i = 0; i < 140; ++i)
    for (int j = 0; j < 140; ++j)
      if (grid[i][j] == 'X') {
        answer += findString(grid, i, j);
      }

  printf("%d\n", answer);

  return EXIT_SUCCESS;
}

int findString(char grid[140][142], int i, int j) {
  int found = 0;

  for (int dx = -1; dx < 2; ++dx)
    for (int dy = -1; dy < 2; ++dy) {
      bool valid = true;

      for (int k = 0; k < 4; ++k) {
        int row = i + dx * k;
        int col = j + dy * k;

        if (row < 0 || row >= 140 || col < 0 || col >= 140 ||
            grid[row][col] != string[k])
          valid = false;
      }

      valid ? ++found : false;
    }

  return found;
}

File: 2024/day-4/part_1.py
STRING: str = "XMAS"


def find_string(grid: list[str], i: int, j: int, row_size: int, col_size: int) -> int:
    found: int = 0

    for dx in range(-1, 2):
        for dy in range(-1, 2):
            valid: bool = True

            for k, char in enumerate(STRING):
                row: int = i + dx * k
                col: int = j + dy * k

                if (
                    row < 0
                    or row >= row_size
                    or col < 0
                    or col >= col_size
                    or grid[row][col] != char
                ):
                    valid = False

            found += valid

    return found


def main() -> None:
    grid: list[str] = []
    answer: int = 0

    with open("input", "r", encoding="UTF-8") as file:
        for line in file:
            grid.append(line.strip("\n"))

    m, n = len(grid), len(grid[0])

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "X":
                answer += find_string(grid, i, j, m, n)

    print(answer)

    return None


if __name__ == "__main__":
    main()
